### PR TITLE
Add an example for failover client

### LINKED ISSUE #19751: Add an examples for webSocket failover client
**Description:**
This is a subtask of #9884

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
Add an example for failover client

Fixes #19751

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: examples/websocket-failover/websocket_failover.bal [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        // Defines the webSocket failover client.
+        http:WebSocketFailoverClient wsClientEp = new (
+        {
+            callbackService: FailoverClientService,
+            // Defines a set of targets.
+            targetUrls: [ "ws://localhost:9095/failover/ws", "ws://localhost:9096/failover/ws",
+            "ws://localhost:9094/failover/ws"],
+            // Failover interval in milliseconds.
+            failoverIntervalInMillis: 2000,
+            // When creating client endpoint, if `readyOnConnect` flag is set to
```
Comment (praneesha) [inline examples/websocket-failover/websocket_failover.bal]:
```suggestion
            // When creating the client endpoint, if the `readyOnConnect` flag is set to
```

File: examples/websocket-failover/websocket_failover.bal [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            // `false` client endpoint does not start reading frames automatically.
```
Comment (praneesha) [inline examples/websocket-failover/websocket_failover.bal]:
```suggestion
            // `false`, the client endpoint does not start reading frames automatically.
```

File: examples/websocket-failover/websocket_failover.bal [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            readyOnConnect: false
+        });
+
+        // Associate connections before starting to read messages.
+        wsClientEp.setAttribute(ASSOCIATED_CONNECTION, caller);
+        caller.setAttribute(ASSOCIATED_CONNECTION, wsClientEp);
+
+        // Once the client is ready to receive frames the remote function `ready`
```
Comment (praneesha) [inline examples/websocket-failover/websocket_failover.bal]:
```suggestion
        // Once the client is ready to receive frames, the remote  `ready` function
```

File: examples/websocket-failover/websocket_failover.bal [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        // of the client need to be called separately.
```
Comment (praneesha) [inline examples/websocket-failover/websocket_failover.bal]:
```suggestion
        // of the client needs to be called separately.
```

File: examples/websocket-failover/websocket_failover.bal [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+        http:WebSocketCaller serverEp = getAssociatedServerEndpoint(caller);
+        var err = serverEp->close(statusCode = statusCode, reason = reason);
+        if (err is http:WebSocketError) {
+            log:printError("Error occurred when closing the connection", err);
+        }
+    }
+};
+
+// Function to retrieve associated client for a particular caller.
```
Comment (praneesha) [inline examples/websocket-failover/websocket_failover.bal]:
```suggestion
// Function to retrieve the associated client for a particular caller.
```

File: examples/websocket-failover/websocket_failover.description [hunk 6]
```diff
@@ -0,0 +1 @@
+// Ballerina users can configure multiple backend in a given failover config. If the client lost the connection by the server being shut down or by getting any IOExceptions, Ballerina automatically try to connect with remaining backends that is specified in the configuration only once.
```
Comment (praneesha) [inline examples/websocket-failover/websocket_failover.description]:
```suggestion
// Ballerina users can configure multiple backends in a given failover config. If the client looses the connection by the server being shut down or by getting any IOExceptions, Ballerina automatically tries to connect with the remaining backends that are specified in the configuration. The trying happens only once for each of the backends.
```

File: examples/websocket-failover/websocket_failover.out [hunk 7]
```diff
@@ -0,0 +1,28 @@
+# To start the services, navigate to the directory that contains the
+# `.bal` file and use the `ballerina run` command.
```
Comment (praneesha) [inline examples/websocket-failover/websocket_failover.out]:
```suggestion
# `.bal` file and run the `ballerina run` command.
```

File: examples/websocket-failover/websocket_failover.out [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+$ ballerina run websocket_service_9095.bal
+[ballerina/http] started HTTP/WS listener 0.0.0.0:9095
+
+$ ballerina run websocket_service_9094.bal
+[ballerina/http] started HTTP/WS listener 0.0.0.0:9094
+
+$ ballerina run websocket_failover.bal
+[ballerina/http] started HTTP/WS listener 0.0.0.0:9090
+
+# Now, this service can be invoked by any WebSocket client using the url "ws://localhost:9090/failover/ws"
```
Comment (praneesha) [inline examples/websocket-failover/websocket_failover.out]:
```suggestion
# Now, this service can be invoked by any WebSocket client using the following URL: "ws://localhost:9090/failover/ws"
```

File: examples/websocket-failover/websocket_failover.out [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+# To check the sample, you can use Chrome or Firefox JavaScript console and run the following commands. <br>
```
Comment (praneesha) [inline examples/websocket-failover/websocket_failover.out]:
```suggestion
# To check the sample, you can use Chrome or the Firefox JavaScript console and run the following commands. <br>
```

File: examples/websocket-failover/websocket_service_9094.bal [hunk 10]
```diff
@@ -0,0 +1,28 @@
+import ballerina/http;
+import ballerina/log;
+
+// Annotations specifying the webSocket service.
```
Comment (praneesha) [inline examples/websocket-failover/websocket_service_9094.bal]:
```suggestion
// Annotations specifying the WebSocket service.
```

File: examples/websocket-failover/websocket_service_9094.bal [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+@http:WebSocketServiceConfig {
+    path: "/failover/ws"
+}
+
+// Define the backend service with port 9094, which is called by the failover client.
+service server on new http:Listener(9094) {
+
+    // This resource gets invoked when a new client connects.
+    // Since messages to the server are not read by the service until the execution of the `onOpen`
+    // resource finishes, operations which should happen before reading messages should be done in
```
Comment (praneesha) [inline examples/websocket-failover/websocket_service_9094.bal]:
```suggestion
    // resource finishes, operations, which should happen before reading the messages should be done in
```

File: examples/websocket-failover/websocket_service_9094.out [hunk 12]
```diff
@@ -0,0 +1,6 @@
+# To start the service, navigate to the directory that contains the
+# `.bal` file, and use the `ballerina run` command.
```
Comment (praneesha) [inline examples/websocket-failover/websocket_service_9094.out]:
```suggestion
# `.bal` file, and run the `ballerina run` command.
```

File: examples/websocket-failover/websocket_service_9094.out [hunk 13]
```diff
@@ -0,0 +1,6 @@
...
+$ ballerina run websocket_service_9094.bal
+
+[ballerina/http] started HTTP/WS listener 0.0.0.0:9094
+2019-11-20 14:05:09,596 INFO  [] - WebSocket client connect wih the server 9094. The Connection ID: 08d40cfffe258b13-00003526-00000001-c69a4c7eb8c373f5-76217d0a
```
Comment (praneesha) [inline examples/websocket-failover/websocket_service_9094.out]:
```suggestion
2019-11-20 14:05:09,596 INFO  [] - WebSocket client connects with the server 9094. The Connection ID: 08d40cfffe258b13-00003526-00000001-c69a4c7eb8c373f5-76217d0a
```

File: examples/websocket-failover/websocket_service_9095.bal [hunk 14]
```diff
@@ -0,0 +1,28 @@
+import ballerina/http;
+import ballerina/log;
+
+// Annotations specifying the webSocket service.
```
Comment (praneesha) [inline examples/websocket-failover/websocket_service_9095.bal]:
```suggestion
// Annotations specifying the WebSocket service.
```

File: examples/websocket-failover/websocket_service_9095.bal [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+@http:WebSocketServiceConfig {
+    path: "/failover/ws"
+}
+
+// Define the backend service with port 9095, which is called by the failover.
+service server on new http:Listener(9095) {
+
+    // This resource gets invoked when a new client connects.
+    // Since messages to the server are not read by the service until the execution of the `onOpen`
+    // resource finishes, operations which should happen before reading messages should be done in
```
Comment (praneesha) [inline examples/websocket-failover/websocket_service_9095.bal]:
```suggestion
    // resource finishes, operations, which should happen before reading messages should be done in
```

File: examples/websocket-failover/websocket_service_9095.out [hunk 16]
```diff
@@ -0,0 +1,6 @@
+# To start the service, navigate to the directory that contains the
+# `.bal` file, and use the `ballerina run` command.
```
Comment (praneesha) [inline examples/websocket-failover/websocket_service_9095.out]:
```suggestion
# `.bal` file, and run the `ballerina run` command.
```

File: examples/websocket-failover/websocket_service_9095.out [hunk 17]
```diff
@@ -0,0 +1,6 @@
...
+$ ballerina run websocket_service_9095.bal
+[ballerina/http] started HTTP/WS listener 0.0.0.0:9095
+2019-11-20 14:04:20,515 INFO  [] - WebSocket client connect wih the server 9095. The Connection ID: 08d40cfffe258b13-00003366-00000001-0cf944d788c2b464-774dc3ff
```
Comment (praneesha) [inline examples/websocket-failover/websocket_service_9095.out]:
```suggestion
2019-11-20 14:04:20,515 INFO  [] - WebSocket client connects with the server 9095. The Connection ID: 08d40cfffe258b13-00003366-00000001-0cf944d788c2b464-774dc3ff
```

**Comment by daneshk** [comment]
Closing this PR temporary. Please reopen once we complete the implemetation

**Pull request merged**